### DISCUSSION TITLE
Is the connection request context maintained for other requests?

### DISCUSSION THREAD

**Discussion by theahura** [miguelgrinberg/Flask-SocketIO#1752] [Q&A] [answered]
According to the implementation notes in the [documentation](https://flask-socketio.readthedocs.io/en/latest/implementation_notes.html#access-to-flask-s-context-globals), Flask-SocketIO allows access to the request context, and this request context is shared for all events that occur over the course of a single connection. 

I have some code like so:
```
@socketio.on('connect')
def connect():
  setattr(request, 'foo', bar)

@socketio.on('baz')
def otherevent():
  assert hasattr(request, 'foo')  # This fails.
```
Does the context not extend to the connection event? Or am I misunderstanding how socketio handles the request context? Or something else?

**Comment by miguelgrinberg** [comment, marked as answer]
You are asking about the request context, but in the code you are modifying the request object and expecting it to be preserved. That's not how it works, the request object is not preserved, each event handler gets its own request object. If you want to store some data for a client you should write it to the user session.

**Comment by theahura** [comment]
Thanks for the (really, incredibly) fast reply, and for the clarification on the object/context. I'm still a bit fuzzy on this distinction -- just to be sure I fully understand, I can do something like:
``` 
@socketio.on('connect')
def connect():
  session['foo'] = bar

@socketio.on('baz')
def otherevent():
  assert 'foo' in session
```
or do I have to do something else to write to the user session?

**Reply by miguelgrinberg** [reply]
This is all you need to do. If you also want the user session changes that you make to be accessible in HTTP routes, then you have to use server-side sessions, as the cookie based sessions that are standard with Flask cannot be updated from the Socket.IO event handler.

**Reply by theahura** [reply]
One last question. The docs say: `If a SocketIO handler modifies the session, the modified session will be preserved for future SocketIO handlers, but regular HTTP route handlers will not see these changes.`

Does this mean flask-socketio maintains the session context across socketio connections from the same machine? How does it do that if it doesnt use cookies? (mostly what I am curious about is if the client can access/edit the session)

**Reply by miguelgrinberg** [reply]
It keeps the session data in memory. But this is only when `manage_session=True`. If you use `manage_session=False` and add a server-side session extension, then everything works and the session can be shared fully between Socket.IO and HTTP handlers.
